File: pkg/types/contract_test.go
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later

package types

import (
	"encoding/json"
	"os"
	"testing"
	"time"

	"github.com/lenaxia/llmsafespaces/pkg/session"
)

// TestGenerateContractFixtures outputs JSON fixtures that the frontend
// contract test validates against. Run with:
//
//	go test -run TestGenerateContractFixtures ./pkg/types/ -v
//
// The output file is consumed by frontend/src/api/contract.test.ts
func TestGenerateContractFixtures(t *testing.T) {
	now := time.Date(2026, 5, 24, 12, 0, 0, 0, time.UTC)
	seenAt := time.Date(2026, 5, 24, 11, 0, 0, 0, time.UTC)
	contextUsed := int64(12500)

	fixtures := map[string]interface{}{
		"AuthConfig": AuthConfig{
			RegistrationEnabled: true,
			OIDCEnabled:         false,
			SSOProviders:        []string{"okta"},
		},
		"ActivateWorkspaceResponse": ActivateWorkspaceResponse{
			Resumed:   "ws-1",
			Suspended: "ws-old",
		},
		"SessionListItem": SessionListItem{
			ID:            "sess-1",
			Title:         "Chat about auth",
			ParentID:      "sess-root",
			LastMessageAt: &now,
			MessageCount:  12,
			Status:        "active",
			LastSeenAt:    &seenAt,
			HasUnread:     true,
			ContextUsed:   &contextUsed,
		},
		"ActiveSessionsResponse": ActiveSessionsResponse{
			Active:    []string{"sess-1", "sess-2"},
			MaxActive: 5,
		},
		"WorkspaceListItem": WorkspaceListItem{
			ID:                "ws-1",
			Name:              "alpha",
			UserID:            "u-123",
			Runtime:           "python:3.11",
			StorageSize:       "5Gi",
			Phase:             "Active",
			MaxActiveSessions: 5,
			CreatedAt:         now,
			UpdatedAt:         now,
		},
		"AuthResponse": AuthResponse{
			Token: "jwt-token",
			User: User{
				ID:       "u-123",
				Username: "alice",
				Email:    "alice@test.com",
				Role:     "user",
				Active:   true,
			},
		},
		// The input fixtures are the platform contract shape
		// (session.InputRequest, #1302) — the same type the REST list
		// routes and the SSE emitter serve. They live in pkg/session
		// because they are the agent-facing contract, not API response
		// types; we include them here because the frontend's
		// contract.test.ts asserts the JSON shape and needs a single
		// fixtures.json to read from.
		"InputRequestQuestion": session.InputRequest{
			ID:            "que_18b28260affeoxXrX1iwPH8wFg",
			SessionID:     "ses_18b28260affeoxXrX1iwPH8wFg",
			RootSessionID: "ses_18b28260affeoxXrX1iwPH8wFg",
			Kind:          session.InputQuestion,
			Question:      "What programming language do you want to use?",
			Header:        "Choose language",
			Options: []session.InputOption{
				{Label: "Go", Description: "Fast compiled language"},
				{Label: "Python", Description: "Easy scripting"},
			},
			Multiple: false,
			Custom:   true,
			Tool:     &session.ToolRef{MessageID: "msg_abc", CallID: "call_xyz"},
		},
		"InputRequestPermission": session.InputRequest{
			ID:            "per_18b28260affeoxXrX1iwPH8wFg",
			SessionID:     "ses_18b28260affeoxXrX1iwPH8wFg",
			RootSessionID: "ses_18b28260affeoxXrX1iwPH8wFg",
			Kind:          session.InputPermission,
			Permission:    "shell",
			Patterns:      []string{"/workspace/src/main.go"},
			Metadata:      map[string]json.RawMessage{"command": []byte(`"go build"`)},
			Always:        []string{"/workspace/*"},
			Tool:          &session.ToolRef{MessageID: "msg_abc", CallID: "call_xyz"},
		},
	}

	data, err := json.MarshalIndent(fixtures, "", "  ")
	if err != nil {
		t.Fatalf("failed to marshal fixtures: %v", err)
	}

	outPath := "../../frontend/src/api/contract-fixtures.json"
	if err := os.WriteFile(outPath, data, 0644); err != nil {
		// If frontend dir doesn't exist (CI without frontend checkout), skip
		t.Skipf("skipping fixture write (frontend not present): %v", err)
	}
	t.Logf("wrote contract fixtures to %s", outPath)
}

File: frontend/src/api/contract.test.ts
import { describe, expect, it } from "vitest";
import fixtures from "./contract-fixtures.json";
import type {
  AuthConfig,
  ActivateWorkspaceResponse,
  ActiveSessionsResponse,
  SessionListItem,
  WorkspaceListItem,
  AuthResponse,
  InputRequest,
} from "./types";

/**
 * Contract tests: validate that the Go-generated JSON fixtures are
 * assignable to our TypeScript types. If a Go struct field is renamed,
 * added, or removed, these tests will catch the drift.
 *
 * Regenerate fixtures: go test -run TestGenerateContractFixtures ./pkg/types/
 */

describe("Go↔TS contract", () => {
  it("AuthConfig matches Go shape", () => {
    const data: AuthConfig = fixtures.AuthConfig;
    expect(data.registrationEnabled).toBe(true);
    expect(data.oidcEnabled).toBe(false);
    expect(data.ssoProviders).toEqual(["okta"]);
  });

  it("ActivateWorkspaceResponse matches Go shape", () => {
    const data: ActivateWorkspaceResponse = fixtures.ActivateWorkspaceResponse;
    expect(data.resumed).toBe("ws-1");
    expect(data.suspended).toBe("ws-old");
  });

  it("ActiveSessionsResponse matches Go shape", () => {
    const data: ActiveSessionsResponse = fixtures.ActiveSessionsResponse;
    expect(data.active).toEqual(["sess-1", "sess-2"]);
    expect(data.maxActive).toBe(5);
  });

  it("SessionListItem matches Go shape", () => {
    const data: SessionListItem = fixtures.SessionListItem;
    expect(data.id).toBe("sess-1");
    expect(data.title).toBe("Chat about auth");
    expect(data.lastMessageAt).toBeDefined();
    expect(data.messageCount).toBe(12);
    expect(data.status).toBe("active");
    expect(data.lastSeenAt).toBeDefined();
    expect(data.hasUnread).toBe(true);
    expect(data.contextUsed).toBe(12500);
  });

  it("WorkspaceListItem matches Go shape", () => {
    const data: WorkspaceListItem = fixtures.WorkspaceListItem;
    expect(data.id).toBe("ws-1");
    expect(data.name).toBe("alpha");
    expect(data.userId).toBe("u-123");
    expect(data.runtime).toBe("python:3.11");
    expect(data.storageSize).toBe("5Gi");
    expect(data.phase).toBe("Active");
    expect(data.maxActiveSessions).toBe(5);
    expect(data.createdAt).toBeDefined();
    expect(data.updatedAt).toBeDefined();
  });

  it("AuthResponse matches Go shape", () => {
    const data: AuthResponse = fixtures.AuthResponse;
    expect(data.token).toBe("jwt-token");
    expect(data.user.id).toBe("u-123");
    expect(data.user.username).toBe("alice");
    expect(data.user.email).toBe("alice@test.com");
    expect(data.user.role).toBe("user");
    expect(data.user.active).toBe(true);
  });

  it("all fixture keys have corresponding tests", () => {
    const testedKeys = [
      "AuthConfig", "ActivateWorkspaceResponse", "ActiveSessionsResponse",
      "SessionListItem", "WorkspaceListItem", "AuthResponse",
      "InputRequestQuestion", "InputRequestPermission",
    ];
    const fixtureKeys = Object.keys(fixtures);
    expect(fixtureKeys.sort()).toEqual(testedKeys.sort());
  });

  it("InputRequest (kind=question) matches Go shape", () => {
    // JSON-module inference widens the kind literal to string; the cast
    // keeps shape-overlap checking, the runtime asserts below carry the
    // value-level drift detection.
    const data = fixtures.InputRequestQuestion as InputRequest;
    expect(data.id).toBe("que_18b28260affeoxXrX1iwPH8wFg");
    expect(data.kind).toBe("question");
    expect(data.sessionId).toBe("ses_18b28260affeoxXrX1iwPH8wFg");
    expect(data.rootSessionId).toBe("ses_18b28260affeoxXrX1iwPH8wFg");
    expect(data.question).toBe("What programming language do you want to use?");
    expect(data.header).toBe("Choose language");
    expect(data.multiple).toBeUndefined(); // omitempty: false is key-absent
    expect(data.custom).toBe(true);
    expect(data.options).toHaveLength(2);
    expect(data.options?.[0]?.label).toBe("Go");
    expect(data.options?.[0]?.description).toBe("Fast compiled language");
    expect(data.tool?.messageId).toBe("msg_abc");
    expect(data.tool?.callId).toBe("call_xyz");
  });

  it("InputRequest (kind=permission) matches Go shape", () => {
    const data = fixtures.InputRequestPermission as InputRequest;
    expect(data.id).toBe("per_18b28260affeoxXrX1iwPH8wFg");
    expect(data.kind).toBe("permission");
    expect(data.sessionId).toBe("ses_18b28260affeoxXrX1iwPH8wFg");
    expect(data.permission).toBe("shell");
    expect(data.patterns).toEqual(["/workspace/src/main.go"]);
    expect(data.metadata).toEqual({ command: "go build" });
    expect(data.always).toEqual(["/workspace/*"]);
    expect(data.tool?.messageId).toBe("msg_abc");
  });
});
